src/tensors/levels/atomiclevel.jl

```diff
@@ -0,0 +1,250 @@
+                if !isnothing(lockVal)
+                    $(lvl.locks)[$idx] = lockVal
+                end
+              end
+          end)
+    assemble_level!(lvl.lvl, ctx, pos_start, pos_stop)
+end
+
+supports_reassembly(lvl::VirtualAtomicLevel) = supports_reassembly(lvl.lvl)
+function reassemble_level!(lvl::VirtualAtomicLevel, ctx, pos_start, pos_stop)
+    pos_start = cache!(ctx, :pos_start, simplify(pos_start, ctx))
+    pos_stop = cache!(ctx, :pos_stop, simplify(pos_stop, ctx))
+    idx = freshen(ctx.code, :idx)
+    lockVal = freshen(ctx.code, :lock)
+    push!(ctx.code.preamble, quote 
+              Finch.resize_if_smaller!($lvl.locks, $(ctx(pos_stop))) 
+              @inbounds for $idx = $(ctx(pos_start)):$(ctx(pos_stop))
+                lockVal = Finch.make_lock(eltype($(lvl.AVal)))
+                if !isnothing(lockVal)
+                    $lvl.locks[$idx] = lockVal
+                end
+              end
+          end)
+    reassemble_level!(lvl.lvl, ctx, pos_start, pos_stop)
+    lvl
+end
+
+function freeze_level!(lvl::VirtualAtomicLevel, ctx, pos)
+    lvl.lvl = freeze_level!(lvl.lvl, ctx, pos)
+    return lvl
+end
+
+function thaw_level!(lvl::VirtualAtomicLevel, ctx::AbstractCompiler, pos)
+    lvl.lvl = thaw_level!(lvl.lvl, ctx, pos)
+    return lvl
+end
+
+function trim_level!(lvl::VirtualAtomicLevel, ctx::AbstractCompiler, pos)
+    # FIXME: Deallocate atomics?
+    posV = ctx(pos)
+    idx = freshen(ctx.code, :idx)
+    push!(ctx.code.preamble, quote
+              resize!($(lvl.locks), $posV)
+          end)
+    lvl.lvl = trim_level!(lvl.lvl, ctx, pos)
+    lvl
+end
+
+function virtual_moveto_level(lvl::VirtualAtomicLevel, ctx::AbstractCompiler, arch)
+    #Add for seperation level too.
+    atomics = freshen(ctx.code, :locksArray)
+
+    push!(ctx.code.preamble, quote
+        $atomics = $(lvl.locks)
+        $(lvl.locks) = $moveto($(lvl.locks), $(ctx(arch)))
+    end)
+    push!(ctx.code.epilogue, quote
+        $(lvl.locks) = $atomics
+    end)
+    virtual_moveto_level(lvl.lvl, ctx, arch)
+end
+
+function instantiate(fbr::VirtualSubFiber{VirtualAtomicLevel}, ctx, mode::Reader, protos)
+    (lvl, pos) = (fbr.lvl, fbr.pos)
+    # lvlp = freshen(ctx.code, lvl.ex, :_lvl)
+    # sym = freshen(ctx.code, lvl.ex, :_after_atomic_lvl)
+    return body = Thunk(
+        body = (ctx) -> begin
+            instantiate(VirtualSubFiber(lvl.lvl, pos), ctx, mode, protos)
+        end,
+    )
+end
+
+function instantiate(fbr::VirtualSubFiber{VirtualAtomicLevel}, ctx, mode::Updater, protos)
+    (lvl, pos) = (fbr.lvl, fbr.pos)
+    sym = freshen(ctx.code, lvl.ex, :after_atomic_lvl)
+    atomicData = freshen(ctx.code, lvl.ex, :atomicArraysAcc)
+    lockVal = freshen(ctx.code, lvl.ex, :lockVal) 
+    dev = lower(virtual_get_device(ctx.code.task), ctx, DefaultStyle())
+    return Thunk(
+        preamble = quote  
+            $atomicData =  promote_val_to_lock($dev, $(lvl.locks), $(ctx(pos)), eltype($(lvl.AVal)))
```

Comment:
could we rename `promote_val_to_lock` to `get_lock` and `get_lock!` to `aquire_lock!`?